File: packages/core/src/showcase/approvalTiers.ts
// F3 — assurance tiers for human approval.
//
// A tier answers ONE question: does an approval carrying this tier count as a
// real human's out-of-band sign-off? The keyring binds only the MAXIMUM tier a
// key may assert. A signed token records the method that was actually used and
// the tier claimed from that method.
//
// Policy picks a FLOOR; an approval satisfies the policy only when its key's
// tier is at or above that floor on the ordered ladder below.

export const AssuranceTier = Object.freeze({
  // An automated / in-session signer. Excluded from every human-sign-off policy.
  UNTRUSTED_AUTOMATION: "untrusted_automation",
  // The old "hardened TTY". Still agent-spoofable (an agent can drive its own
  // PTY), so it is explicitly NOT trusted for human sign-off. Weak fallback only.
  SAME_CHANNEL_OPERATOR_CONFIRMATION: "same_channel_operator_confirmation",
  // A token signed by a key held in host/OS custody outside the agent's scope,
  // keyring-verified. THE v1 target tier for "a human really approved this".
  TRUSTED_HOST_USER_PRESENCE: "trusted_host_user_presence"
  // (deferred slot: webauthn_hardware — a hardware-attested tier above this.)
} as const);

export type AssuranceTier = (typeof AssuranceTier)[keyof typeof AssuranceTier];

export const AssuranceMethod = Object.freeze({
  AUTOMATION: "automation",
  SAME_CHANNEL: "same_channel",
  OS_PRESENCE: "os_presence"
} as const);

export type AssuranceMethod = (typeof AssuranceMethod)[keyof typeof AssuranceMethod];

export const ASSURANCE_METHOD_TO_TIER: Record<AssuranceMethod, AssuranceTier> = Object.freeze({
  [AssuranceMethod.AUTOMATION]: AssuranceTier.UNTRUSTED_AUTOMATION,
  [AssuranceMethod.SAME_CHANNEL]: AssuranceTier.SAME_CHANNEL_OPERATOR_CONFIRMATION,
  [AssuranceMethod.OS_PRESENCE]: AssuranceTier.TRUSTED_HOST_USER_PRESENCE
});

// Ordered weakest -> strongest. Index is the ladder rank; a higher (deferred)
// hardware tier would append here.
const LADDER: AssuranceTier[] = [
  AssuranceTier.UNTRUSTED_AUTOMATION,
  AssuranceTier.SAME_CHANNEL_OPERATOR_CONFIRMATION,
  AssuranceTier.TRUSTED_HOST_USER_PRESENCE
];

// Fail-closed: anything unrecognised (including undefined / a bogus string) is
// treated as the weakest tier.
export function normalizeAssuranceTier(value: unknown): AssuranceTier {
  return LADDER.find((tier) => tier === value) ?? AssuranceTier.UNTRUSTED_AUTOMATION;
}

export function isAssuranceTier(value: unknown): value is AssuranceTier {
  return LADDER.some((tier) => tier === value);
}

export function isAssuranceMethod(value: unknown): value is AssuranceMethod {
  return Object.values(AssuranceMethod).some((method) => method === value);
}

export function assuranceTierForMethod(method: AssuranceMethod): AssuranceTier {
  return ASSURANCE_METHOD_TO_TIER[method];
}

// Is `tier` at or above `floor` on the ladder?
export function tierMeetsFloor(tier: AssuranceTier, floor: AssuranceTier): boolean {
  return LADDER.indexOf(tier) >= LADDER.indexOf(floor);
}

// Does a tier count as a genuine human sign-off (v1 target and above)?
export function trustedForHumanSignoff(tier: AssuranceTier): boolean {
  return tierMeetsFloor(tier, AssuranceTier.TRUSTED_HOST_USER_PRESENCE);
}

File: UseCasesCore/Sources/UseCasesCore/Showcase/ApprovalTiers.swift
/// How strongly an approval proves a real human signed it off
/// (showcase/approvalTiers.ts), ordered weakest to strongest.
public enum AssuranceTier: String, CaseIterable, Comparable, Sendable {
  /// An automated or in-session signer; never a human sign-off.
  case untrustedAutomation = "untrusted_automation"
  /// An operator confirming on the agent's own channel; agent-spoofable.
  case sameChannelOperatorConfirmation = "same_channel_operator_confirmation"
  /// A keyring-verified key held in host custody outside the agent's scope.
  case trustedHostUserPresence = "trusted_host_user_presence"
  /// A WebAuthn assertion verified against a pinned hardware credential.
  case webAuthnHardware = "webauthn_hardware"

  /// The ladder rank, weakest first.
  private var rank: Int {
    Self.allCases.firstIndex(of: self) ?? 0
  }

  public static func < (
    left: AssuranceTier,
    right: AssuranceTier,
  ) -> Bool {
    left.rank < right.rank
  }

  /// `normalizeAssuranceTier`: anything unrecognised, absent included, is the
  /// weakest tier.
  static func normalized(_ value: JSONValue?) -> AssuranceTier {
    value?.stringValue.flatMap(recognised) ?? .untrustedAutomation
  }

  /// `normalizeAssuranceTier` for a keyring cap.
  static func normalized(_ tier: KeyringAssuranceTier?) -> AssuranceTier {
    let cap = tier.flatMap { tier in
      AssuranceTier(rawValue: tier.rawValue)
    }
    return cap ?? .untrustedAutomation
  }

  /// `isAssuranceTier`: a string spelling exactly one of the tiers.
  static func recognised(_ text: String) -> AssuranceTier? {
    allCases.first { tier in
      JavaScriptString.identical(tier.rawValue, text)
    }
  }

  /// `tierMeetsFloor`.
  func meets(_ floor: AssuranceTier) -> Bool {
    self >= floor
  }

  /// `trustedForHumanSignoff`: at or above host user presence.
  public var isTrustedForHumanSignOff: Bool {
    meets(.trustedHostUserPresence)
  }
}

/// How an approval was actually captured; each fixes the tier it may claim.
public enum AssuranceMethod: String, CaseIterable, Sendable {
  case automation
  case sameChannel = "same_channel"
  case operatingSystemPresence = "os_presence"
  case webAuthn = "webauthn"

  /// `ASSURANCE_METHOD_TO_TIER`.
  public var tier: AssuranceTier {
    switch self {
    case .automation: .untrustedAutomation
    case .sameChannel: .sameChannelOperatorConfirmation
    case .operatingSystemPresence: .trustedHostUserPresence
    case .webAuthn: .webAuthnHardware
    }
  }

  /// `isAssuranceMethod`.
  static func recognised(_ value: JSONValue?) -> AssuranceMethod? {
    guard let text = value?.stringValue else {
      return nil
    }
    return allCases.first { method in
      JavaScriptString.identical(method.rawValue, text)
    }
  }
}

/// The approval floor a plan's policies set (showcase/approvalPolicy.ts).
enum ApprovalPolicy {
  /// `approvalAssuranceFloorForPlan`: the strongest `minimum_assurance_tier`
  /// among the policies that require a user, each defaulting to host user
  /// presence; host user presence when none requires one.
  static func assuranceFloor(forPlan plan: JSONValue?) throws(ShowcaseError) -> AssuranceTier {
    var floors: [AssuranceTier] = []
    let items = try ShowcaseJavaScript.arrayOrEmpty(
      ShowcaseJavaScript.optionalMember(plan, "selected_items"),
      calling: "map",
    )
    var policies: [JSONValue?] = []
    for item in items {
      try policies.append(ShowcaseJavaScript.member(item, "approval_policy_snapshot"))
    }
    for policy in policies {
      guard try requiresUserApproval(policy) else {
        continue
      }
      let declared = ShowcaseJavaScript.optionalMember(policy, "minimum_assurance_tier")
      floors
        .append(declared?.stringValue.flatMap(AssuranceTier.recognised) ?? .trustedHostUserPresence)
    }
    guard var floor = floors.first else {
      return .trustedHostUserPresence
    }
    for candidate in floors.dropFirst() where candidate.meets(floor) {
      floor = candidate
    }
    return floor
  }

  /// `policyRequiresUserApproval`: a predefined policy with an array of
  /// requirements, one of them an object whose approver is a user.
  private static func requiresUserApproval(_ policy: JSONValue?) throws(ShowcaseError) -> Bool {
    guard try ShowcaseJavaScript.isString(ShowcaseJavaScript.member(policy, "mode"), "predefined"),
          let requirements = ShowcaseJavaScript.optionalMember(policy, "requirements")?.arrayValue
    else {
      return false
    }
    return requirements.contains { requirement in
      ShowcaseJavaScript.isString(requirement.objectValue?["approver_type"], "user")
    }
  }

  /// The `planRequiresUserApproval` / `userApprovalRequired` check both
  /// appendShowcaseEvent.ts and replayRun.ts spell out: unlike the floor's
  /// check, a `null` requirement is read, and throws.
  static func requiresUserApproval(plan: JSONValue?) throws(ShowcaseError) -> Bool {
    let items = try ShowcaseJavaScript.arrayOrEmpty(
      ShowcaseJavaScript.optionalMember(plan, "selected_items"),
      calling: "some",
    )
    for item in items {
      let policy = try ShowcaseJavaScript.member(item, "approval_policy_snapshot")
      guard try ShowcaseJavaScript.isString(
        ShowcaseJavaScript.member(policy, "mode"),
        "predefined",
      ),
        let requirements = ShowcaseJavaScript.optionalMember(policy, "requirements")?.arrayValue
      else {
        continue
      }
      for requirement in requirements {
        let approver = try ShowcaseJavaScript.member(requirement, "approver_type")
        if ShowcaseJavaScript.isString(approver, "user") {
          return true
        }
      }
    }
    return false
  }
}
